lib/harmony/harmonytask/harmonytask.go

```diff
@@ -0,0 +1,462 @@
+
+type TaskTypeDetails struct {
+	// Max returns how many tasks this machine can run of this type.
+	// Negative means unrestricted.
+	Max int
+
+	// Name is the task name to be added to the task list.
+	Name string
+
+	// Peak costs to Do() the task.
+	Cost resources.Resources
+
+	// Max Failure count before the job is dropped.
+	// 0 = retry forever
+	MaxFailures uint
+
+	// Follow another task's completion via this task's creation.
+	// The function should populate extraInfo from data
+	// available from the previous task's tables, using the given TaskID.
+	// It should also return success if the trigger succeeded.
+	// NOTE: if refatoring tasks, see if your task is
+	// necessary. Ex: Is the sector state correct for your stage to run?
+	Follows map[string]func(TaskID, AddTaskFunc) bool
+}
+
+// TaskInterface must be implemented in order to have a task used by harmonytask.
+type TaskInterface interface {
+	// Do the task assigned. Call stillOwned before making single-writer-only
+	// changes to ensure the work has not been stolen.
+	// This is the ONLY function that should attempt to do the work, and must
+	// ONLY be called by harmonytask.
+	// Indicate if the task no-longer needs scheduling with done=true including
+	// cases where it's past the deadline.
+	Do(taskID TaskID, stillOwned func() bool) (done bool, err error)
+
+	// CanAccept should return if the task can run on this machine. It should
+	// return null if the task type is not allowed on this machine.
+	// It should select the task it most wants to accomplish.
+	CanAccept([]TaskID) (*TaskID, error)
+
+	// TypeDetails() returns static details about how this task behaves and
+	// how this machine will run it. Read once at the beginning.
+	TypeDetails() TaskTypeDetails
+
+	// This listener will consume all external sources continuously for work.
+	// Do() may also be called from a backlog of work. This must not
+	// start doing the work (it still must be scheduled).
+	// Note: Task de-duplication should happen in ExtraInfoFunc by
+	//  returning false, typically by determining from the tx that the work
+	//  exists already. The easy way is to have a unique joint index
+	//  across all fields that will be common.
+	// Adder should typically only add its own task type, but multiple
+	//   is possible for when 1 trigger starts 2 things.
+	// Usage Example:
+	// func (b *BazType)Adder(addTask AddTaskFunc) {
+	//	  for {
+	//      bazMaker := <- bazChannel
+	//	    addTask("baz", func(t harmonytask.TaskID, txn db.Transaction) bool {
+	//	       _, err := txn.Exec(`INSERT INTO bazInfoTable (taskID, qix, mot)
+	//			  VALUES ($1,$2,$3)`, id, bazMaker.qix, bazMaker.mot)
+	//         if err != nil {
+	//				scream(err)
+	//	 		 	return false
+	//		   }
+	// 		   return true
+	//		})
+	//	  }
+	// }
+	Adder(AddTaskFunc)
+}
+
+type AddTaskFunc func(extraInfo func(TaskID, *harmonydb.Tx) bool)
+
+type TaskEngine struct {
+	ctx            context.Context
```

Comment:
Can we avoid the context in the struct?

Reply:
Oh, I see, this is used for graceful exit.

Typically we use two `chan struct{}` for that, much more robust, and plays nicely with uber/fx.Lifecycle, which makes sure that everything actually stops cleanly, and in the correct order.

Basically you have 2 channels:
* `stopping`, which is closed when we want to stop
* `stopped`, which is closed in the goroutine that we want to exit
* Then you have a `Stop(shutdownCtx)` method, which closes `stopping`, and waits for `stopped` (or `shutdownCtx`, which is a context provided by fx.Lifecycle, and has a 15s timeout iirc)

Reply:
The advantage of using context is libraries like PGX respond to this and will actually cancel DB operations in progress in order to return. This will cause API call receivers during this period to error at the DB calls so that no new work could get picked-up.
I like hooking it up to fx.Lifecycle so only 1 call is needed to close down everything
	lc.Append(fx.Hook{
		OnStop: func(ctx context.Context) error {
		       e.grace()
                        e.gracefulShutdown()
		},
	})
, but I see a lot of value in keeping the context too (for the DB benefit). WDYT?

Reply:
> will actually cancel DB operations in progress in order to return

I don't think we have any long-running DB operations that we'd want to cancel in-flight, I'd expect harmonytask to stop cleanly first, and only after that the PG connection to close